### PR TITLE
[COREV] Define patterns for the generation of ALU instructions

### PR DESCRIPTION
**Pull request by serkm**

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 1]
```diff
@@ -431,57 +429,165 @@ def macu   : PatFrag<(ops node:$rd, node:$rs1, node:$rs2),
...
+def betweenu : PatFrags<(ops node:$upperBound, node:$value),
+                        [(umin (umax node:$value, 0), node:$upperBound),
+                         (umax (umin node:$value, node:$upperBound), 0)]>;
+
+
+def clip : PatFrag<(ops node:$upperBound, node:$value),
+                   (between negativePowerOf2, node:$upperBound, node:$value), [{
+
+  if (auto bound1 = dyn_cast<ConstantSDNode>(N->getOperand(0)->getOperand(1))) {
+    if (auto bound2 = dyn_cast<ConstantSDNode>(N->getOperand(1))) {
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td:445]: Can you add comments, what operands you're trying to get here?

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 2]
```diff
@@ -431,57 +429,165 @@ def macu   : PatFrag<(ops node:$rd, node:$rs1, node:$rs2),
...
+  def : Pat<(shiftRound (sub GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_SUBRN GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+  def : Pat<(ushiftRound (sub GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_SUBURN GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+
+  def : Pat<(sra (add GPR:$rd, GPR:$rs1), GPR:$rs2),
+            (CV_ADDNR GPR:$rd, GPR:$rs1, GPR:$rs2)>;
+  def : Pat<(srl (add GPR:$rd, GPR:$rs1), GPR:$rs2),
+            (CV_ADDUNR GPR:$rd, GPR:$rs1, GPR:$rs2)>;
+  def : Pat<(sra (add (add GPR:$rd, GPR:$rs1), (roundBit GPR:$rs2)), GPR:$rs2),
```
Comment (flip1995) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td:569]: Nothing prevents the `$rs2` from being `>=32`. Should `roundBit` include a `and GPR:$rs2,  0x1f`?

Comment (serkm) [reply llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td:569]: IIUC the shift operation is [undefined](https://llvm.org/docs/LangRef.html#id148) on shift amounts `>=32`, so we don't need to take care of that case.

Comment (flip1995) [reply llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td:569]: TIL. Yeah, makes sense, so nothing to do here.

**Comment by flip1995** [comment]
LGTM. @jeremybennett

**Comment by jeremybennett** [comment]
Apologies - I merged #13 before this one. Could you please rebase and I'll merge.

**Comment by serkm** [comment]
It's rebased. @jeremybennett

**Comment by jeremybennett** [comment]
Thanks. Good to merge.

**Pull request merged**